src/crimson/osd/heartbeat.cc

```diff
@@ -320,23 +320,25 @@ seastar::future<> Heartbeat::maybe_share_osdmap(
   }
 
   if (osdmap_epoch <= peer.get_last_epoch_sent()) {
-    logger().info("{} latest epoch sent {} is already later "
-                  "than osdmap epoch of {}",
-                  __func__ , peer.get_last_epoch_sent(),
-                  osdmap_epoch);
     return seastar::now();
   }
 
-  logger().info("{} peer id: {} epoch is {} while osdmap is {}",
-                __func__ , from, m->map_epoch, osdmap_epoch);
-  if (osdmap_epoch > m->map_epoch) {
-    logger().debug("{} sharing osdmap epoch of {} with peer id {}",
-                   __func__, osdmap_epoch, from);
-    // Peer's newest map is m->map_epoch. Therfore it misses
-    // the osdmaps in the range of `m->map_epoch` to `osdmap_epoch`.
-    return service.send_incremental_map_to_osd(from, m->map_epoch);
-  }
-  return seastar::now();
+  assert(osdmap_epoch > peer_epoch);
+  logger().debug("{} sharing osdmap epoch of {} with peer id {}",
+                 __func__, osdmap_epoch, from);
+  // Peer's newest map is m->map_epoch. Therfore it misses
+  // the osdmaps in the range of `m->map_epoch` to `osdmap_epoch`.
+  return service.send_incremental_map_to_osd(from, m->map_epoch
```

Comment:
should be service.send_incremental_map_to_osd(from, peer.get_last_epoch_sent())

Reply:
Hmm, I think we actually want to update last_epoch_sent prior to sending it.  The delivery mechanism is reliable -- it will get there unless the OSD in question is marked down.  We don't want to queue another message containing the same map.

Reply:
> should be service.send_incremental_map_to_osd(from, peer.get_last_epoch_sent())

Ack

> I think we actually want to update last_epoch_sent prior to sending it.

We are updating `last_epoch_sent` prior to sending and after sending.

-------

I further simplified the changes, local tests look good.